### PR TITLE
ThemeProvider: updated to use abstract renderers, defaulting to merge-styles but overridable

### PR DESCRIPTION
**Pull request by dzearing**
This update does a few things:

* ThemeProvider updated to new composition model
* ThemeProvider styling now done using `makeClasses` helper
* Registering styles now done through an abstraction, provided through the props, but also through rescaffolding the ThemeProvider with a new default renderer using the `useThemeProvider` component factory hook.
* `makeClasses` and `makeStyles` have been updated to utilize the renderer from the theme.

### CODE REVIEW

File: packages/react-button/src/components/Button/__snapshots__/Button.test.tsx.snap [hunk 1]
```diff
@@ -1,212 +1,19 @@
...
->
-  <span
-    className=
-
-        {
-          line-height: 1;
-        }
-  >
+<button>
+  <span>
```
Comment (xugao) [inline packages/react-button/src/components/Button/__snapshots__/Button.test.tsx.snap:5]: i'm confused why this change is here

File: change/@fluentui-react-theme-provider-2020-08-26-12-10-14-chore-evaluate-cssinjs-cost.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Minor tweaks in theme styling.",
```
Comment (xugao) [inline change/@fluentui-react-theme-provider-2020-08-26-12-10-14-chore-evaluate-cssinjs-cost.json:3]:
```suggestion
  "type": "minor",
  "comment": "Support passing different renderer and default to use merge-styles",
```

File: packages/react-theme-provider/src/useThemeProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+) => {
+  const state: ThemeProviderState = mergeProps(
+    {
+      // eslint-disable-next-line react-hooks/rules-of-hooks
+      ref: ref || React.useRef(),
+      as: 'div',
+    },
+    defaultProps,
+    props,
+  ) as ThemeProviderState;
```
Comment (xugao) [inline packages/react-theme-provider/src/useThemeProvider.tsx:24]: is the casting needed?

File: packages/react-theme-provider/src/useThemeProviderClasses.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { makeClasses } from './makeClasses';
+
+export const useThemeProviderClasses = makeClasses({
+  root: {
```
Comment (xugao) [inline packages/react-theme-provider/src/useThemeProviderClasses.tsx:4]: please update styles and add comments based on latest ThemeProvider scss

File: packages/react-theme-provider/src/makeClasses.test.tsx [hunk 5]
```diff
@@ -3,21 +3,35 @@ import * as React from 'react';
 import { makeClasses } from './makeClasses';
 import { safeMount } from '@uifabric/test-utilities';
 import { Stylesheet, InjectionMode } from '@uifabric/merge-styles';
+import { mergeStylesRenderer } from './styleRenderers/mergeStylesRenderer';
+import { StyleRendererContext } from './styleRenderers/useStyleRenderer';
+
+const MergeStylesProvider = ({ children }: React.PropsWithChildren<{}>) => (
+  <StyleRendererContext.Provider value={mergeStylesRenderer}>{children}</StyleRendererContext.Provider>
+);
```
Comment (xugao) [inline packages/react-theme-provider/src/makeClasses.test.tsx:11]: move this to a test util and use everywhere

File: packages/react-theme-provider/src/useThemeTokenClass.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import * as React from 'react';
+import { ThemeProviderState } from './ThemeProvider.types';
+import { tokensToStyleObject } from './tokensToStyleObject';
+
+/**
+ * Give the tokens defined within the theme, create a className
+ * to append to the className on state. Memoizes the result to
+ * avoid re-computations.
+ */
+export const useThemeTokenClass = (state: ThemeProviderState) => {
```
Comment (xugao) [inline packages/react-theme-provider/src/useThemeTokenClass.ts:11]: call this `useThemeProviderClass` instead?

File: packages/react-theme-provider/src/makeStyles.ts [hunk 7]
```diff
@@ -35,32 +39,33 @@ const graphSet = (graphNode: Map<any, any>, path: any[], value: any) => {
...
 
-  return () => {
+  return (theme?: Theme, renderer?: StyleRenderer) => {
     const win = useWindow();
-    const theme = useCustomizationSettings(['theme']).theme;
-    const path = theme ? [_seed, win, theme] : [_seed, win];
+
+    // Expected: theme is either always provided or never.
+    theme = theme || useTheme();
+    renderer = (renderer || useStyleRenderer()) as StyleRenderer;
```
Comment (xugao) [inline packages/react-theme-provider/src/makeStyles.ts:59]:
```suggestion
    renderer = (renderer || useStyleRenderer());
```

is `as StyleRenderer` needed here?

Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts:59]: only in TS 3.7. In 4.0.2, no. :-/

Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts:59]: quite confusing.

File: packages/react-theme-provider/src/useThemeTokenClass.ts (hunk 6 above)
Comment (dzearing) [reply packages/react-theme-provider/src/useThemeTokenClass.ts:11]: This is specifically for the tokens in theme; not the `ThemeProvider` classes, which is separate and uses `makeClasses`.

File: packages/react-button/src/components/Button/__snapshots__/Button.test.tsx.snap (hunk 1 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/__snapshots__/Button.test.tsx.snap:5]: well that's confusing. I will try and figure this out.

File: packages/react-theme-provider/src/makeStyles.ts [hunk 8]
```diff
@@ -35,32 +39,33 @@ const graphSet = (graphNode: Map<any, any>, path: any[], value: any) => {
...
+
+    const id = renderer.getId();
+    const path = theme ? [id, win, theme] : [id, win];
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts]: I would like to keep the key list as short as possible. I've adjusted the logic a bit though, theme will only be an entry for styles which depend on the theme via function.

File: packages/react-theme-provider/src/makeStyles.ts [hunk 9]
```diff
@@ -35,32 +39,33 @@ const graphSet = (graphNode: Map<any, any>, path: any[], value: any) => {
...
   // Create graph of inputs to map to output.
   const graph = new Map();
 
-  return () => {
+  return (theme?: Theme, renderer?: StyleRenderer) => {
     const win = useWindow();
-    const theme = useCustomizationSettings(['theme']).theme;
-    const path = theme ? [_seed, win, theme] : [_seed, win];
+
+    // Expected: theme is either always provided or never.
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts]: fixed.

File: packages/react-theme-provider/src/makeStyles.ts [hunk 10]
```diff
@@ -35,32 +39,33 @@ const graphSet = (graphNode: Map<any, any>, path: any[], value: any) => {
...
+
+    const id = renderer.getId();
+    const path = theme ? [id, win, theme] : [id, win];
     let value = graphGet(graph, path);
 
     if (!value) {
       const styles = typeof styleOrFunction !== 'function' ? styleOrFunction : styleOrFunction(theme);
-      value = mergeStyleSets(styles);
+
+      value = renderer.renderStyles(styles, { targetWindow: win!, rtl: !!theme.rtl });
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts]: fixed.

File: packages/react-theme-provider/src/useThemeProvider.tsx (hunk 3 above)
Comment (dzearing) [reply packages/react-theme-provider/src/useThemeProvider.tsx:24]: I've adjusted mergeProps to take in the state as a generic.

**Comment by dzearing** [comment]
Need to investigate bundle size before merging.

File: packages/react-button/src/components/Button/Button.test.tsx [hunk 11]
```diff
@@ -1,9 +1,18 @@
...
 import { isConformant } from '@fluentui/react-conformance';
-import { Button } from './Button';
+import { Button as ButtonComponent } from './Button';
 import * as renderer from 'react-test-renderer';
 import { mount, ReactWrapper } from 'enzyme';
+import { MergeStylesProvider } from '@fluentui/react-theme-provider';
+import { ButtonProps } from './Button.types';
+
+/** Use merge-styles provider to ensure styles show up in snapshots. */
+const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
```
Comment (dzearing) [inline packages/react-button/src/components/Button/Button.test.tsx:11]: @xugao This was kind of annoying but it's the only way to ensure styles are rendered. Do you have any other ideas here to make this easier?

File: packages/react-theme-provider/src/__snapshots__/ThemeProvider.test.tsx.snap [hunk 12]
```diff
@@ -2,1853 +2,1933 @@
...
+
+      {
+        background: var(--body-background);
+        box-sizing: border-box;
+        color: var(--body-contentColor);
+        font-family: var(--body-fontFamily);
+        font-size: var(--body-fontSize);
+        font-weight: var(--body-fontWeight);
+        line-height: var(--body-fontLineHeight);
+      }
```
Comment (dzearing) [inline packages/react-theme-provider/src/__snapshots__/ThemeProvider.test.tsx.snap:15]: These shouldn't be here by default

File: packages/react-theme-provider/src/useThemeTokenClass.ts (hunk 6 above)
Comment (dzearing) [reply packages/react-theme-provider/src/useThemeTokenClass.ts:11]: hmmm, actually, this could move into `useThemeProviderClass`

Comment (dzearing) [reply packages/react-theme-provider/src/useThemeTokenClass.ts]: resolved.

File: packages/react-button/src/components/Button/Button.test.tsx (hunk 11 above)
Comment (xugao) [reply packages/react-button/src/components/Button/Button.test.tsx:11]: i can't think anything better than this: https://stackoverflow.com/questions/59017016/is-it-possible-to-wrap-each-individual-test-in-a-wrapper-component

File: packages/merge-styles/src/transforms/provideUnits.ts [hunk 13]
```diff
@@ -12,11 +12,12 @@ const NON_PIXEL_NUMBER_PROPS = [
 ];
 
 export function provideUnits(rulePairs: (string | number)[], index: number): void {
-  const name = rulePairs[index];
+  const name = rulePairs[index] as string;
   const value = rulePairs[index + 1];
 
   if (typeof value === 'number') {
-    const unit = NON_PIXEL_NUMBER_PROPS.indexOf(name as string) === -1 ? 'px' : '';
+    const convertToPixel = NON_PIXEL_NUMBER_PROPS.indexOf(name as string) === -1 && name.indexOf('--') === -1;
```
Comment (xugao) [inline packages/merge-styles/src/transforms/provideUnits.ts:19]: i think we should create a `isCssVariable(name: string): boolean` helper in merge-styles and use it here. most likely will be handy later as well

File: change/@uifabric-merge-styles-2020-09-10-14-38-58-chore-theme-renderers.json [hunk 14]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "minor",
+  "comment": "Exporting additional typings. ",
```
Comment (xugao) [inline change/@uifabric-merge-styles-2020-09-10-14-38-58-chore-theme-renderers.json:3]:
```suggestion
  "comment": "Exporting additional typings; fix css variable value unit.",
```

File: packages/react-button/package.json [hunk 15]
```diff
@@ -59,13 +59,13 @@
...
-    "office-ui-fabric-react": "^7.137.1",
     "tslib": "^1.10.0"
   },
   "peerDependencies": {
     "@types/react": ">=16.8.0 <17.0.0",
     "@types/react-dom": ">=16.8.0 <17.0.0",
     "react": ">=16.8.0 <17.0.0",
-    "react-dom": ">=16.8.0 <17.0.0"
+    "react-dom": ">=16.8.0 <17.0.0",
+    "office-ui-fabric-react": ">=7.137.1 <8.0.0"
```
Comment (xugao) [inline packages/react-button/package.json:69]: could you add a new issue to track reverting this change later?

File: packages/react-compose/src/next/makeMergeProps.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+              if (Array.isArray(propValue)) {
+                // for arrays, replace.
+                target[propName] = propValue;
+              } else {
+                target[propName] = target[propName] || {};
+
+                if (
+                  typeof target[propName] !== 'object' ||
+                  // React.isValidElement(propValue) ||
+                  // (propValue && typeof propValue === 'object' && propValue.hasOwnProperty('current')) ||
```
Comment (xugao) [inline packages/react-compose/src/next/makeMergeProps.ts:40]: remove

File: packages/react-theme-provider/src/makeStyles.ts [hunk 17]
```diff
@@ -35,32 +39,34 @@ const graphSet = (graphNode: Map<any, any>, path: any[], value: any) => {
...
   const graph = new Map();
 
-  return () => {
+  return (theme?: Theme, renderer?: StyleRenderer) => {
     const win = useWindow();
-    const theme = useCustomizationSettings(['theme']).theme;
-    const path = theme ? [_seed, win, theme] : [_seed, win];
+
+    // Expected: theme and renderer are either always provided or never.
+    theme = theme || (useCustomizationSettings(['theme']).theme as Theme) || {};
```
Comment (xugao) [inline packages/react-theme-provider/src/makeStyles.ts:58]: let's just fix `useTheme` so we can `useTheme` here? it should be very straight-forward:
```
/**
 * React hook for programatically accessing the theme.
 */
export const useTheme = (): Theme => {
  const theme = useContext(ThemeContext);
  const legacyTheme = useCompatTheme();
  if (theme) {
    return theme;
  }

  return legacyTheme;
};
```

File: packages/react-theme-provider/src/useThemeProviderClasses.tsx [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  return {
+    root: [
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      tokensToStyleObject(tokens) as any,
+      {
+        color: 'var(--body-contentColor)',
+        fontFamily: 'var(--body-fontFamily)',
+        fontWeight: 'var(--body-fontWeight)',
+        fontSize: 'var(--body-fontSize)',
+        lineHeight: 'var(--body-fontLineHeight)',
```
Comment (xugao) [inline packages/react-theme-provider/src/useThemeProviderClasses.tsx:16]: pls remove lineHeight as well

File: packages/merge-styles/src/transforms/provideUnits.ts (hunk 13 above)
Comment (dzearing) [reply packages/merge-styles/src/transforms/provideUnits.ts:19]: `--webkit-foo` is not a css variable but would match here as well. `isCssVariableOrVendorPrefixed` ?

Comment (xugao) [reply packages/merge-styles/src/transforms/provideUnits.ts:19]: ah ic. either that or do `const isCssVariableOrVendorPrefixed = name.indexOf('--') === -1` here for how. goal is to make this piece code more understandable

File: packages/react-compose/src/next/makeMergeProps.ts (hunk 16 above)
Comment (dzearing) [reply packages/react-compose/src/next/makeMergeProps.ts]: added back.

File: packages/react-button/package.json (hunk 15 above)
Comment (dzearing) [reply packages/react-button/package.json:69]: #15028

File: packages/react-theme-provider/src/useThemeProviderClasses.tsx (hunk 18 above)
Comment (dzearing) [reply packages/react-theme-provider/src/useThemeProviderClasses.tsx:16]: Done.

File: packages/react-button/package.json (hunk 15 above)
Comment (dzearing) [reply packages/react-button/package.json:69]: #15028

File: packages/react-theme-provider/src/makeStyles.ts (hunk 17 above)
Comment (dzearing) [reply packages/react-theme-provider/src/makeStyles.ts]: Done.

File: packages/merge-styles/src/transforms/provideUnits.ts (hunk 13 above)
Comment (dzearing) [reply packages/merge-styles/src/transforms/provideUnits.ts]: Updated code to be more clear, good suggestion.

**Review by xugao — APPROVED** [review]
LGTM. Only thing is that the api file changes still look incorrect to me - line ending should not be changed
